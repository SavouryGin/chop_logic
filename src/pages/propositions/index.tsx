import React from 'react';
import TabList from 'components/tab-list';
import formatClassName from 'helpers/formatters/format-class-name';
import DirectProofsEditor from './elements/direct-proofs-editor';
import NaturalProofsEditor from './elements/natural-proofs-editor';
import DefinitionsOfPropositions from './elements/definitions-of-propositions';
import HeuristicsOfPropositions from './elements/heuristics-of-propositions';
import RulesOfInference from './elements/rules-of-inference';
import { TabItem } from 'types';
import { Icon } from 'enums';
import { pageTitle, tabTitles } from 'assets/texts/propositions';
import { useAppSelector } from 'hooks';
import { settingsSelectors } from 'store/settings/selectors';

import './styles.scss';

export const propositionsTabs: TabItem[] = [
  {
    tabContent: <DirectProofsEditor />,
    tabTitle: tabTitles.direct,
    tabId: 'direct-proofs-editor',
  },
  {
    tabContent: <NaturalProofsEditor />,
    tabTitle: tabTitles.natural,
    tabId: 'natural-proofs-editor',
  },
  {
    tabContent: <DefinitionsOfPropositions />,
    tabTitle: tabTitles.definitions,
    tabId: 'definitions-of-propositions',
  },
  {
    tabContent: <HeuristicsOfPropositions />,
    tabTitle: tabTitles.heuristics,
    tabId: 'heuristics-of-propositions',
  },
  {
    tabContent: <RulesOfInference />,
    tabTitle: tabTitles.rules,
    tabId: 'rules-of-inference',
  },
];

function Propositions(): React.ReactElement {
  const language = useAppSelector(settingsSelectors.getLanguage);
  return (
    <article className='propositions'>
      <h2 className={formatClassName(['propositions__title', Icon.Propositions])}>{pageTitle[language]}</h2>
      <TabList tabs={propositionsTabs} className='propositions__tabs' />
    </article>
  );
}

export default Propositions;
