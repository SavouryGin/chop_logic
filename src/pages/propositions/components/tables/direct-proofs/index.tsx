import React, { memo, useEffect, useState } from 'react';
import Table from 'components/table';
import constants from 'pages/propositions/constants';
import texts from 'texts/propositions/elements';
import { dpActions as actions } from 'store/propositions/direct-proofs';
import { dpSelectors as selectors } from 'store/propositions/direct-proofs/selectors';
import { settingsSelectors } from 'store/settings/selectors';
import { useAppDispatch, useAppSelector } from 'hooks';

const DirectProofsEditorTable = (): React.ReactElement => {
  const dispatch = useAppDispatch();
  const tableData = useAppSelector(selectors.tableData);
  const language = useAppSelector(settingsSelectors.language);
  const [selectedIds, setSelectedIds] = useState<string[]>(useAppSelector(selectors.selectedIds));
  const takeSelectedIds = (ids: string[]) => setSelectedIds(ids);

  useEffect(() => {
    dispatch(actions.setSelectedIds(selectedIds));
  }, [selectedIds]);

  const noStepsFiller = <div className='direct-proofs-editor__filler'>{texts.fillerText[language]}</div>;

  return (
    <div className='direct-proofs-editor__table'>
      <Table
        columns={constants.directProofsEditorTableColumns}
        data={tableData}
        hasCheckboxColumn
        passSelectedIds={takeSelectedIds}
        id='propositions-dp-editor-table'
      />
      {tableData.length === 0 && noStepsFiller}
    </div>
  );
};

export default memo(DirectProofsEditorTable);
