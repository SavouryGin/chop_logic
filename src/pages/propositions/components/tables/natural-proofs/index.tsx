import React, { useEffect, useState } from 'react';
import Table from 'components/table';
import constants from 'pages/propositions/constants';
import propositionsElementsTexts from 'texts/propositions/elements';
import { propositionsNPActions as actions } from 'store/propositions/natural-proofs';
import { npSelectors as selectors } from 'store/propositions/natural-proofs/selectors';
import { settingsSelectors } from 'store/settings/selectors';
import { useAppDispatch, useAppSelector } from 'hooks';

const NaturalProofsEditorTable = (): React.ReactElement => {
  const dispatch = useAppDispatch();
  const language = useAppSelector(settingsSelectors.language);
  const tableData = useAppSelector(selectors.tableData);
  const [selectedIds, setSelectedIds] = useState<string[]>(useAppSelector(selectors.selectedIds));
  const takeSelectedIds = (ids: string[]) => setSelectedIds(ids);

  useEffect(() => {
    dispatch(actions.setSelectedIds(selectedIds));
  }, [selectedIds]);

  const noStepsFiller = <div className='natural-proofs-editor__filler'>{propositionsElementsTexts.fillerNaturalText[language]}</div>;

  return (
    <div className='natural-proofs-editor__table'>
      <Table
        columns={constants.naturalProofsEditorTableColumns}
        data={tableData}
        hasCheckboxColumn
        passSelectedIds={takeSelectedIds}
        id='propositions-np-editor-table'
      />
      {tableData.length === 0 && noStepsFiller}
    </div>
  );
};

export default NaturalProofsEditorTable;
