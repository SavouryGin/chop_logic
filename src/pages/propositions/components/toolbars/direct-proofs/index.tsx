import Button from 'components/controls/button';
import React, { memo } from 'react';
import { ButtonID } from 'enums';
import { propositionsDPActions as actions } from 'store/propositions/direct-proofs/slice';
import { propositionsDPSelectors as selectors } from 'store/propositions/direct-proofs/selectors';
import { soundPlayer } from 'helpers/sounds';
import { useAppDispatch, useAppSelector, useIsImplicationEliminationPossible } from 'hooks';

const DirectProofsEditorToolbar = () => {
  const dispatch = useAppDispatch();
  const selectedIds = useAppSelector(selectors.getSelectedIds);
  const tableDataLength = useAppSelector(selectors.getTableDataLength);
  const isReiterationDisabled = selectedIds.length !== 1;
  const isDeleteDisabled = selectedIds.length === 0;
  const isImplicationEliminationEnabled = useIsImplicationEliminationPossible(selectedIds);
  const isReplacerDisabled = tableDataLength === 0;
  const selectedItems = useAppSelector(selectors.getSelectedTableItems);

  const deleteSteps = () => {
    dispatch(actions.deleteSteps({ isConfirmed: false }));
  };

  const reiterateStep = () => {
    dispatch(actions.reiterateStep());
  };

  const openPremise = () => {
    dispatch(actions.setUpFlag({ flag: 'isPremiseOpened', value: true }));
  };

  const openIC = () => {
    dispatch(actions.setUpFlag({ flag: 'isImplicationCreationOpened', value: true }));
  };

  const openID = () => {
    dispatch(actions.setUpFlag({ flag: 'isImplicationDistributionOpened', value: true }));
  };

  const openCR = () => {
    dispatch(actions.setUpFlag({ flag: 'isContradictionRealizationOpened', value: true }));
  };

  const performIE = () => {
    dispatch(actions.eliminateImplication(selectedItems));
  };

  const openReplacer = () => {
    dispatch(actions.setUpFlag({ flag: 'isReplacerFormOpened', value: true }));
  };

  return (
    <div className='direct-proofs-editor__toolbar'>
      <Button buttonId={ButtonID.Premise} sound={soundPlayer.keyboard} size='large' onClick={openPremise} />
      <Button
        buttonId={ButtonID.Reiteration}
        sound={soundPlayer.keyboard}
        size='large'
        isDisabled={isReiterationDisabled}
        onClick={reiterateStep}
      />
      <Button
        buttonId={ButtonID.Replace}
        sound={soundPlayer.keyboard}
        size='large'
        onClick={openReplacer}
        isDisabled={isReplacerDisabled}
      />
      <Button buttonId={ButtonID.Delete} sound={soundPlayer.keyboard} size='large' isDisabled={isDeleteDisabled} onClick={deleteSteps} />
      <Button buttonId={ButtonID.ImplicationCreation} sound={soundPlayer.slideClick} size='large' onClick={openIC} />
      <Button buttonId={ButtonID.ImplicationDistribution} sound={soundPlayer.slideClick} size='large' onClick={openID} />
      <Button buttonId={ButtonID.ContradictionRealization} sound={soundPlayer.slideClick} size='large' onClick={openCR} />
      <Button
        buttonId={ButtonID.ImplicationElimination}
        sound={soundPlayer.slideClick}
        isDisabled={!isImplicationEliminationEnabled}
        size='large'
        onClick={performIE}
      />
    </div>
  );
};

export default memo(DirectProofsEditorToolbar);
