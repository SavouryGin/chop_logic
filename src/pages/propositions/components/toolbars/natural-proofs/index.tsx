import Button from 'components/controls/button';
import React from 'react';
import { ButtonID } from 'enums';
import { npActions as actions } from 'store/propositions/natural-proofs';
import { npSelectors as selectors } from 'store/propositions/natural-proofs/selectors';
import { soundPlayer } from 'helpers/sounds';
import {
  useAppDispatch,
  useAppSelector,
  useIsAndEliminationPossible,
  useIsEquivalenceEliminationPossible,
  useIsEquivalenceIntroductionPossible,
  useIsImpliesEliminationForNPPossible,
  useIsImpliesIntroductionPossible,
  useIsNotEliminationPossible,
  useIsNotIntroductionPossible,
  useIsOrEliminationPossible,
  useIsPremisePossible,
} from 'hooks';

const NaturalProofsEditorToolbar = (): React.ReactElement => {
  const dispatch = useAppDispatch();
  const isPremiseDisabled = !useIsPremisePossible();
  const selectedIds = useAppSelector(selectors.selectedIds);
  const tableDataLength = useAppSelector(selectors.tableDataLength);
  const isReplacerDisabled = tableDataLength === 0;
  const isReiterationDisabled = selectedIds.length !== 1;
  const isDeleteDisabled = selectedIds.length === 0;
  const isOrIntroductionDisabled = selectedIds.length === 0;
  const isOrEliminationDisabled = !useIsOrEliminationPossible(selectedIds);
  const isAndIntroductionDisabled = !selectedIds.length;
  const isAndEliminationDisabled = !useIsAndEliminationPossible(selectedIds);
  const isNotIntroductionDisabled = !useIsNotIntroductionPossible(selectedIds);
  const isNotEliminationDisabled = !useIsNotEliminationPossible(selectedIds);
  const isEquivIntroductionDisabled = !useIsEquivalenceIntroductionPossible(selectedIds);
  const isEquivEliminationDisabled = !useIsEquivalenceEliminationPossible(selectedIds);
  const isImpliesEliminationDisabled = !useIsImpliesEliminationForNPPossible(selectedIds);
  const isImpliesIntroductionDisabled = !useIsImpliesIntroductionPossible(selectedIds);

  const openPremise = () => {
    dispatch(actions.setUpFlag({ flag: 'isPremiseOpened', value: true }));
  };

  const openAssumption = () => {
    dispatch(actions.setUpFlag({ flag: 'isAssumptionOpened', value: true }));
  };

  const deleteSteps = () => {
    dispatch(actions.deleteSteps({ isConfirmed: false }));
  };

  const reiterateStep = () => {
    dispatch(actions.reiterateStep());
  };

  const openOrIntroduction = () => {
    dispatch(actions.setUpFlag({ flag: 'isOrIntroductionFormOpened', value: true }));
  };

  const eliminateDisjunction = () => {
    dispatch(actions.eliminateDisjunction());
  };

  const createConjunction = () => {
    dispatch(actions.createConjunction());
  };

  const eliminateConjunction = () => {
    dispatch(actions.eliminateConjunction());
  };

  const createNegation = () => {
    dispatch(actions.createNegation());
  };

  const eliminateNegation = () => {
    dispatch(actions.eliminateNegation());
  };

  const createEquivalence = () => {
    dispatch(actions.createEquivalence());
  };

  const eliminateEquivalence = () => {
    dispatch(actions.eliminateEquivalence());
  };

  const eliminateImplication = () => {
    dispatch(actions.eliminateImplication());
  };

  const createImplication = () => {
    dispatch(actions.createImplication());
  };

  const openShortcut = () => {
    dispatch(actions.setUpFlag({ flag: 'isShortcutOpened', value: true }));
  };

  const openReplacer = () => {
    dispatch(actions.setUpFlag({ flag: 'isReplacerFormOpened', value: true }));
  };

  return (
    <div className='natural-proofs-editor__toolbar'>
      <Button buttonId={ButtonID.Premise} sound={soundPlayer.keyboard} view='large' onClick={openPremise} isDisabled={isPremiseDisabled} />
      <Button
        buttonId={ButtonID.Reiteration}
        sound={soundPlayer.keyboard}
        view='large'
        onClick={reiterateStep}
        isDisabled={isReiterationDisabled}
      />
      <Button
        buttonId={ButtonID.Replace}
        sound={soundPlayer.keyboard}
        view='large'
        onClick={openReplacer}
        isDisabled={isReplacerDisabled}
      />
      <Button
        buttonId={ButtonID.NotIntroduction}
        sound={soundPlayer.keyboard}
        view='large'
        onClick={createNegation}
        isDisabled={isNotIntroductionDisabled}
      />
      <Button
        buttonId={ButtonID.AndIntroduction}
        sound={soundPlayer.keyboard}
        view='large'
        onClick={createConjunction}
        isDisabled={isAndIntroductionDisabled}
      />
      <Button
        buttonId={ButtonID.OrIntroduction}
        sound={soundPlayer.keyboard}
        view='large'
        onClick={openOrIntroduction}
        isDisabled={isOrIntroductionDisabled}
      />
      <Button
        buttonId={ButtonID.ImpliesIntroduction}
        sound={soundPlayer.keyboard}
        view='large'
        onClick={createImplication}
        isDisabled={isImpliesIntroductionDisabled}
      />
      <Button
        buttonId={ButtonID.EquivIntroduction}
        sound={soundPlayer.keyboard}
        view='large'
        isDisabled={isEquivIntroductionDisabled}
        onClick={createEquivalence}
      />
      <Button buttonId={ButtonID.Assumption} sound={soundPlayer.keyboard} view='large' onClick={openAssumption} />
      <Button buttonId={ButtonID.Delete} sound={soundPlayer.keyboard} view='large' onClick={deleteSteps} isDisabled={isDeleteDisabled} />
      <Button buttonId={ButtonID.Shortcut} sound={soundPlayer.keyboard} view='large' onClick={openShortcut} />
      <Button
        buttonId={ButtonID.NotElimination}
        sound={soundPlayer.keyboard}
        view='large'
        onClick={eliminateNegation}
        isDisabled={isNotEliminationDisabled}
      />
      <Button
        buttonId={ButtonID.AndElimination}
        sound={soundPlayer.keyboard}
        view='large'
        onClick={eliminateConjunction}
        isDisabled={isAndEliminationDisabled}
      />
      <Button
        buttonId={ButtonID.OrElimination}
        sound={soundPlayer.keyboard}
        view='large'
        onClick={eliminateDisjunction}
        isDisabled={isOrEliminationDisabled}
      />
      <Button
        buttonId={ButtonID.ImpliesElimination}
        sound={soundPlayer.keyboard}
        view='large'
        onClick={eliminateImplication}
        isDisabled={isImpliesEliminationDisabled}
      />
      <Button
        buttonId={ButtonID.EquivElimination}
        sound={soundPlayer.keyboard}
        view='large'
        onClick={eliminateEquivalence}
        isDisabled={isEquivEliminationDisabled}
      />
    </div>
  );
};

export default NaturalProofsEditorToolbar;
