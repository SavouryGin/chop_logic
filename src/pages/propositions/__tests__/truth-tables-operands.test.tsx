import React from 'react';
import TruthTablesOperands from '../tabs/truth-tables/operands';
import renderWithRedux from 'utils/testing/render-with-redux';
import { DP_INITIAL_STATE } from 'store/propositions/direct-proofs/initial-state';
import { NP_INITIAL_STATE } from 'store/propositions/natural-proofs/initial-state';
import { combineReducers } from '@reduxjs/toolkit';
import { propositionsDPSlice } from 'store/propositions/direct-proofs';
import { propositionsNPSlice } from 'store/propositions/natural-proofs';
import { settingsInitialState, settingsSlice } from 'store/settings';

describe('TruthTablesOperands tests:', () => {
  const mockedReducer = combineReducers({
    settings: settingsSlice.reducer,
    propositionsDP: propositionsDPSlice.reducer,
    propositionsNP: propositionsNPSlice.reducer,
  });

  const mockedState = {
    settings: settingsInitialState,
    propositionsDP: DP_INITIAL_STATE,
    propositionsNP: NP_INITIAL_STATE,
  };

  it('should match the snapshot', () => {
    const { asFragment } = renderWithRedux(<TruthTablesOperands />, mockedReducer, mockedState);
    expect(asFragment()).toMatchSnapshot();
  });
});
