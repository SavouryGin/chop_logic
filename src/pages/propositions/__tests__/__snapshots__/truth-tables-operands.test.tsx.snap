// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TruthTablesOperands tests: should match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="truth-tables_operands"
  >
    <table
      class="table"
    >
      <thead
        class="table__head"
      >
        <tr
          class="table__row"
        >
          <th
            class="table__cell table__cell_heading"
          >
            <span>
              <span>
                <span
                  class="katex"
                >
                  <span
                    class="katex-mathml"
                  >
                    <math>
                      <semantics>
                        <mrow>
                          <mi
                            mathvariant="normal"
                          >
                            ⊥
                          </mi>
                        </mrow>
                        <annotation
                          encoding="application/x-tex"
                        >
                          \\bot
                        </annotation>
                      </semantics>
                    </math>
                  </span>
                  <span
                    aria-hidden="true"
                    class="katex-html"
                  >
                    <span
                      class="base"
                    >
                      <span
                        class="strut"
                        style="height:0.69444em;vertical-align:0em;"
                      />
                      <span
                        class="mord"
                      >
                        ⊥
                      </span>
                    </span>
                  </span>
                </span>
              </span>
            </span>
          </th>
          <th
            class="table__cell table__cell_heading"
          >
            <span>
              <span>
                <span
                  class="katex"
                >
                  <span
                    class="katex-mathml"
                  >
                    <math>
                      <semantics>
                        <mrow>
                          <mo>
                            ∧
                          </mo>
                        </mrow>
                        <annotation
                          encoding="application/x-tex"
                        >
                          \\land
                        </annotation>
                      </semantics>
                    </math>
                  </span>
                  <span
                    aria-hidden="true"
                    class="katex-html"
                  >
                    <span
                      class="base"
                    >
                      <span
                        class="strut"
                        style="height:0.55556em;vertical-align:0em;"
                      />
                      <span
                        class="mord"
                      >
                        ∧
                      </span>
                    </span>
                  </span>
                </span>
              </span>
            </span>
          </th>
          <th
            class="table__cell table__cell_heading"
          >
            <span>
              <span>
                <span
                  class="katex"
                >
                  <span
                    class="katex-mathml"
                  >
                    <math>
                      <semantics>
                        <mrow>
                          <mo>
                            ↛
                          </mo>
                        </mrow>
                        <annotation
                          encoding="application/x-tex"
                        >
                          \\nrightarrow
                        </annotation>
                      </semantics>
                    </math>
                  </span>
                  <span
                    aria-hidden="true"
                    class="katex-html"
                  >
                    <span
                      class="base"
                    >
                      <span
                        class="strut"
                        style="height:0.36687em;vertical-align:0em;"
                      />
                      <span
                        class="mrel amsrm"
                      >
                        ↛
                      </span>
                    </span>
                  </span>
                </span>
              </span>
            </span>
          </th>
          <th
            class="table__cell table__cell_heading"
          >
            <span>
              <span>
                <span
                  class="katex"
                >
                  <span
                    class="katex-mathml"
                  >
                    <math>
                      <semantics>
                        <mrow>
                          <mi>
                            p
                          </mi>
                        </mrow>
                        <annotation
                          encoding="application/x-tex"
                        >
                          p
                        </annotation>
                      </semantics>
                    </math>
                  </span>
                  <span
                    aria-hidden="true"
                    class="katex-html"
                  >
                    <span
                      class="base"
                    >
                      <span
                        class="strut"
                        style="height:0.625em;vertical-align:-0.19444em;"
                      />
                      <span
                        class="mord mathdefault"
                      >
                        p
                      </span>
                    </span>
                  </span>
                </span>
              </span>
            </span>
          </th>
          <th
            class="table__cell table__cell_heading"
          >
            <span>
              <span>
                <span
                  class="katex"
                >
                  <span
                    class="katex-mathml"
                  >
                    <math>
                      <semantics>
                        <mrow>
                          <mo>
                            ↚
                          </mo>
                        </mrow>
                        <annotation
                          encoding="application/x-tex"
                        >
                          \\nleftarrow
                        </annotation>
                      </semantics>
                    </math>
                  </span>
                  <span
                    aria-hidden="true"
                    class="katex-html"
                  >
                    <span
                      class="base"
                    >
                      <span
                        class="strut"
                        style="height:0.36687em;vertical-align:0em;"
                      />
                      <span
                        class="mrel amsrm"
                      >
                        ↚
                      </span>
                    </span>
                  </span>
                </span>
              </span>
            </span>
          </th>
          <th
            class="table__cell table__cell_heading"
          >
            <span>
              <span>
                <span
                  class="katex"
                >
                  <span
                    class="katex-mathml"
                  >
                    <math>
                      <semantics>
                        <mrow>
                          <mi>
                            q
                          </mi>
                        </mrow>
                        <annotation
                          encoding="application/x-tex"
                        >
                          q
                        </annotation>
                      </semantics>
                    </math>
                  </span>
                  <span
                    aria-hidden="true"
                    class="katex-html"
                  >
                    <span
                      class="base"
                    >
                      <span
                        class="strut"
                        style="height:0.625em;vertical-align:-0.19444em;"
                      />
                      <span
                        class="mord mathdefault"
                        style="margin-right:0.03588em;"
                      >
                        q
                      </span>
                    </span>
                  </span>
                </span>
              </span>
            </span>
          </th>
          <th
            class="table__cell table__cell_heading"
          >
            <span>
              <span>
                <span
                  class="katex"
                >
                  <span
                    class="katex-mathml"
                  >
                    <math>
                      <semantics>
                        <mrow>
                          <mo>
                            ⊕
                          </mo>
                        </mrow>
                        <annotation
                          encoding="application/x-tex"
                        >
                          \\oplus
                        </annotation>
                      </semantics>
                    </math>
                  </span>
                  <span
                    aria-hidden="true"
                    class="katex-html"
                  >
                    <span
                      class="base"
                    >
                      <span
                        class="strut"
                        style="height:0.66666em;vertical-align:-0.08333em;"
                      />
                      <span
                        class="mord"
                      >
                        ⊕
                      </span>
                    </span>
                  </span>
                </span>
              </span>
            </span>
          </th>
          <th
            class="table__cell table__cell_heading"
          >
            <span>
              <span>
                <span
                  class="katex"
                >
                  <span
                    class="katex-mathml"
                  >
                    <math>
                      <semantics>
                        <mrow>
                          <mo>
                            ∨
                          </mo>
                        </mrow>
                        <annotation
                          encoding="application/x-tex"
                        >
                          \\lor
                        </annotation>
                      </semantics>
                    </math>
                  </span>
                  <span
                    aria-hidden="true"
                    class="katex-html"
                  >
                    <span
                      class="base"
                    >
                      <span
                        class="strut"
                        style="height:0.55556em;vertical-align:0em;"
                      />
                      <span
                        class="mord"
                      >
                        ∨
                      </span>
                    </span>
                  </span>
                </span>
              </span>
            </span>
          </th>
          <th
            class="table__cell table__cell_heading"
          >
            <span>
              <span>
                <span
                  class="katex"
                >
                  <span
                    class="katex-mathml"
                  >
                    <math>
                      <semantics>
                        <mrow>
                          <mo>
                            ↓
                          </mo>
                        </mrow>
                        <annotation
                          encoding="application/x-tex"
                        >
                          \\downarrow
                        </annotation>
                      </semantics>
                    </math>
                  </span>
                  <span
                    aria-hidden="true"
                    class="katex-html"
                  >
                    <span
                      class="base"
                    >
                      <span
                        class="strut"
                        style="height:0.8888799999999999em;vertical-align:-0.19444em;"
                      />
                      <span
                        class="mrel"
                      >
                        ↓
                      </span>
                    </span>
                  </span>
                </span>
              </span>
            </span>
          </th>
          <th
            class="table__cell table__cell_heading"
          >
            <span>
              <span>
                <span
                  class="katex"
                >
                  <span
                    class="katex-mathml"
                  >
                    <math>
                      <semantics>
                        <mrow>
                          <mo>
                            ↔
                          </mo>
                        </mrow>
                        <annotation
                          encoding="application/x-tex"
                        >
                          \\leftrightarrow
                        </annotation>
                      </semantics>
                    </math>
                  </span>
                  <span
                    aria-hidden="true"
                    class="katex-html"
                  >
                    <span
                      class="base"
                    >
                      <span
                        class="strut"
                        style="height:0.36687em;vertical-align:0em;"
                      />
                      <span
                        class="mrel"
                      >
                        ↔
                      </span>
                    </span>
                  </span>
                </span>
              </span>
            </span>
          </th>
          <th
            class="table__cell table__cell_heading"
          >
            <span>
              <span>
                <span
                  class="katex"
                >
                  <span
                    class="katex-mathml"
                  >
                    <math>
                      <semantics>
                        <mrow>
                          <mi
                            mathvariant="normal"
                          >
                            ¬
                          </mi>
                          <mi>
                            q
                          </mi>
                        </mrow>
                        <annotation
                          encoding="application/x-tex"
                        >
                          \\neg q
                        </annotation>
                      </semantics>
                    </math>
                  </span>
                  <span
                    aria-hidden="true"
                    class="katex-html"
                  >
                    <span
                      class="base"
                    >
                      <span
                        class="strut"
                        style="height:0.625em;vertical-align:-0.19444em;"
                      />
                      <span
                        class="mord"
                      >
                        ¬
                      </span>
                      <span
                        class="mord mathdefault"
                        style="margin-right:0.03588em;"
                      >
                        q
                      </span>
                    </span>
                  </span>
                </span>
              </span>
            </span>
          </th>
          <th
            class="table__cell table__cell_heading"
          >
            <span>
              <span>
                <span
                  class="katex"
                >
                  <span
                    class="katex-mathml"
                  >
                    <math>
                      <semantics>
                        <mrow>
                          <mo>
                            ←
                          </mo>
                        </mrow>
                        <annotation
                          encoding="application/x-tex"
                        >
                          \\leftarrow
                        </annotation>
                      </semantics>
                    </math>
                  </span>
                  <span
                    aria-hidden="true"
                    class="katex-html"
                  >
                    <span
                      class="base"
                    >
                      <span
                        class="strut"
                        style="height:0.36687em;vertical-align:0em;"
                      />
                      <span
                        class="mrel"
                      >
                        ←
                      </span>
                    </span>
                  </span>
                </span>
              </span>
            </span>
          </th>
          <th
            class="table__cell table__cell_heading"
          >
            <span>
              <span>
                <span
                  class="katex"
                >
                  <span
                    class="katex-mathml"
                  >
                    <math>
                      <semantics>
                        <mrow>
                          <mi
                            mathvariant="normal"
                          >
                            ¬
                          </mi>
                          <mi>
                            p
                          </mi>
                        </mrow>
                        <annotation
                          encoding="application/x-tex"
                        >
                          \\neg p
                        </annotation>
                      </semantics>
                    </math>
                  </span>
                  <span
                    aria-hidden="true"
                    class="katex-html"
                  >
                    <span
                      class="base"
                    >
                      <span
                        class="strut"
                        style="height:0.625em;vertical-align:-0.19444em;"
                      />
                      <span
                        class="mord"
                      >
                        ¬
                      </span>
                      <span
                        class="mord mathdefault"
                      >
                        p
                      </span>
                    </span>
                  </span>
                </span>
              </span>
            </span>
          </th>
          <th
            class="table__cell table__cell_heading"
          >
            <span>
              <span>
                <span
                  class="katex"
                >
                  <span
                    class="katex-mathml"
                  >
                    <math>
                      <semantics>
                        <mrow>
                          <mo>
                            →
                          </mo>
                        </mrow>
                        <annotation
                          encoding="application/x-tex"
                        >
                          \\rightarrow
                        </annotation>
                      </semantics>
                    </math>
                  </span>
                  <span
                    aria-hidden="true"
                    class="katex-html"
                  >
                    <span
                      class="base"
                    >
                      <span
                        class="strut"
                        style="height:0.36687em;vertical-align:0em;"
                      />
                      <span
                        class="mrel"
                      >
                        →
                      </span>
                    </span>
                  </span>
                </span>
              </span>
            </span>
          </th>
          <th
            class="table__cell table__cell_heading"
          >
            <span>
              <span>
                <span
                  class="katex"
                >
                  <span
                    class="katex-mathml"
                  >
                    <math>
                      <semantics>
                        <mrow>
                          <mi
                            mathvariant="normal"
                          >
                            ∣
                          </mi>
                        </mrow>
                        <annotation
                          encoding="application/x-tex"
                        >
                          |
                        </annotation>
                      </semantics>
                    </math>
                  </span>
                  <span
                    aria-hidden="true"
                    class="katex-html"
                  >
                    <span
                      class="base"
                    >
                      <span
                        class="strut"
                        style="height:1em;vertical-align:-0.25em;"
                      />
                      <span
                        class="mord"
                      >
                        ∣
                      </span>
                    </span>
                  </span>
                </span>
              </span>
            </span>
          </th>
          <th
            class="table__cell table__cell_heading"
          >
            <span>
              <span>
                <span
                  class="katex"
                >
                  <span
                    class="katex-mathml"
                  >
                    <math>
                      <semantics>
                        <mrow>
                          <mi
                            mathvariant="normal"
                          >
                            ⊥
                          </mi>
                        </mrow>
                        <annotation
                          encoding="application/x-tex"
                        >
                          \\bot
                        </annotation>
                      </semantics>
                    </math>
                  </span>
                  <span
                    aria-hidden="true"
                    class="katex-html"
                  >
                    <span
                      class="base"
                    >
                      <span
                        class="strut"
                        style="height:0.69444em;vertical-align:0em;"
                      />
                      <span
                        class="mord"
                      >
                        ⊥
                      </span>
                    </span>
                  </span>
                </span>
              </span>
            </span>
          </th>
        </tr>
      </thead>
      <tbody
        class="table__body"
      >
        <tr
          class="table__row"
        >
          <td
            class="table__cell"
          >
            0
          </td>
          <td
            class="table__cell"
          >
            0
          </td>
          <td
            class="table__cell"
          >
            0
          </td>
          <td
            class="table__cell"
          >
            0
          </td>
          <td
            class="table__cell"
          >
            0
          </td>
          <td
            class="table__cell"
          >
            0
          </td>
          <td
            class="table__cell"
          >
            0
          </td>
          <td
            class="table__cell"
          >
            0
          </td>
          <td
            class="table__cell"
          >
            1
          </td>
          <td
            class="table__cell"
          >
            1
          </td>
          <td
            class="table__cell"
          >
            1
          </td>
          <td
            class="table__cell"
          >
            1
          </td>
          <td
            class="table__cell"
          >
            1
          </td>
          <td
            class="table__cell"
          >
            1
          </td>
          <td
            class="table__cell"
          >
            1
          </td>
          <td
            class="table__cell"
          >
            1
          </td>
        </tr>
        <tr
          class="table__row"
        >
          <td
            class="table__cell"
          >
            0
          </td>
          <td
            class="table__cell"
          >
            0
          </td>
          <td
            class="table__cell"
          >
            0
          </td>
          <td
            class="table__cell"
          >
            0
          </td>
          <td
            class="table__cell"
          >
            1
          </td>
          <td
            class="table__cell"
          >
            1
          </td>
          <td
            class="table__cell"
          >
            1
          </td>
          <td
            class="table__cell"
          >
            1
          </td>
          <td
            class="table__cell"
          >
            0
          </td>
          <td
            class="table__cell"
          >
            0
          </td>
          <td
            class="table__cell"
          >
            0
          </td>
          <td
            class="table__cell"
          >
            0
          </td>
          <td
            class="table__cell"
          >
            1
          </td>
          <td
            class="table__cell"
          >
            1
          </td>
          <td
            class="table__cell"
          >
            1
          </td>
          <td
            class="table__cell"
          >
            1
          </td>
        </tr>
        <tr
          class="table__row"
        >
          <td
            class="table__cell"
          >
            0
          </td>
          <td
            class="table__cell"
          >
            0
          </td>
          <td
            class="table__cell"
          >
            1
          </td>
          <td
            class="table__cell"
          >
            1
          </td>
          <td
            class="table__cell"
          >
            0
          </td>
          <td
            class="table__cell"
          >
            0
          </td>
          <td
            class="table__cell"
          >
            1
          </td>
          <td
            class="table__cell"
          >
            1
          </td>
          <td
            class="table__cell"
          >
            0
          </td>
          <td
            class="table__cell"
          >
            0
          </td>
          <td
            class="table__cell"
          >
            1
          </td>
          <td
            class="table__cell"
          >
            1
          </td>
          <td
            class="table__cell"
          >
            0
          </td>
          <td
            class="table__cell"
          >
            0
          </td>
          <td
            class="table__cell"
          >
            1
          </td>
          <td
            class="table__cell"
          >
            1
          </td>
        </tr>
        <tr
          class="table__row"
        >
          <td
            class="table__cell"
          >
            0
          </td>
          <td
            class="table__cell"
          >
            1
          </td>
          <td
            class="table__cell"
          >
            0
          </td>
          <td
            class="table__cell"
          >
            1
          </td>
          <td
            class="table__cell"
          >
            0
          </td>
          <td
            class="table__cell"
          >
            1
          </td>
          <td
            class="table__cell"
          >
            0
          </td>
          <td
            class="table__cell"
          >
            1
          </td>
          <td
            class="table__cell"
          >
            0
          </td>
          <td
            class="table__cell"
          >
            1
          </td>
          <td
            class="table__cell"
          >
            0
          </td>
          <td
            class="table__cell"
          >
            1
          </td>
          <td
            class="table__cell"
          >
            0
          </td>
          <td
            class="table__cell"
          >
            1
          </td>
          <td
            class="table__cell"
          >
            0
          </td>
          <td
            class="table__cell"
          >
            1
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</DocumentFragment>
`;
